File: C++/graph_based_algorithms/src/breadth_first_search.cpp
#include "pathfinding_utils.h"

int bfs(std::vector<std::vector<char>>& grid)
{
    // get start and end positions
    std::pair<int, int> start, end;
    std::tie(start, end) = findLocations(grid);
    
    // Initialize the queue
    std::queue<std::pair<std::pair<int, int>, int>> queue; // (position, distance)
    queue.push({start, 0});

    // Track visited cells
    std::set<std::pair<int, int>> visited; 
    visited.insert(start);
    
    std::map<std::pair<int, int>, std::pair<int, int>> came_from; // slower than unsing std::unordered_map
    came_from[start] = {-1, -1};

    while (!queue.empty()) {
        std::pair<std::pair<int, int>, int> currentPosition = queue.front();
        queue.pop();

        // check if end is reached
        int distance = currentPosition.second;
        if (currentPosition.first == end) {
            
            markPath(grid, came_from, start, end);
            return distance; // Return the distance to the end position
        }

        // Explore neighbors
        for (const std::pair<int, int>& direction: directions) {

            int x = currentPosition.first.first + direction.first;
            int y = currentPosition.first.second + direction.second;

            // Check if the next cell is traversable    
            if (isTraversable(grid, visited, x, y)) {
                visited.insert({x, y});
                queue.push({{x, y}, distance + 1});
                came_from[{x, y}] = currentPosition.first;
            }
        }
    }
    return -1; // End is unreachable
}

File: python/graph_based_algorithms/breadth_first_search.py
from collections import deque

def bfs(grid, start, end):
    """
    Perform breadth-first search (BFS) on a 2D grid to find the shortest path from start to end.

    Args:
        grid (List[List[str]]): The 2D grid representing the map.
        start (Tuple[int, int]): Coordinates of the starting point.
        end (Tuple[int, int]): Coordinates of the ending point.

    Returns:
        Tuple[List[List[str]], int]: A tuple containing:
            - The grid with the path marked.
            - The number of steps in the shortest path, or -1 if the end is unreachable.
    """
    # Define possible moves (left, right, down, up)
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    non_traversable = {'-', '|'}
    
    # Initialize the BFS queue
    queue = deque([(start, 0)])  # (position, distance)
    visited = {start}            # Track visited cells
    came_from = {start: None}     # Track the path to each cell
    
    # Clone the grid to avoid modifying the original during path marking
    path_grid = [row[:] for row in grid]
    rows, cols = len(grid), len(grid[0])

    def is_traversable(x, y):
        """Check if a cell is within bounds, traversable, and not yet visited."""
        return 0 <= x < rows and 0 <= y < cols and grid[x][y] not in non_traversable and (x, y) not in visited

    # BFS to explore paths
    while queue:
        (x, y), dist = queue.popleft()
        
        # Check if we've reached the end
        if (x, y) == end:
            # Backtrack to mark the path
            current = end
            while current != start:
                cx, cy = current
                path_grid[cx][cy] = '*'  # Mark each cell along the path
                current = came_from[current]
            return path_grid, dist
        
        # Explore neighbors
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if is_traversable(nx, ny):
                visited.add((nx, ny))
                queue.append(((nx, ny), dist + 1))
                came_from[(nx, ny)] = (x, y)  # Record the previous cell

    # If end is unreachable
    return path_grid, -1
